### PR TITLE
net/eui64: add eui64_to_eui48()

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

Provide helper function to mangle a EUI-64 into an EUI-48. (Like `eui48_to_eui64()`)
This can be useful to get predictable, locally unique addreses.

### Issues/PRs references

https://github.com/RIOT-OS/RIOT/pull/15083#issuecomment-698744777

### CODE REVIEW

File: sys/include/net/eui48.h [hunk 1]
```diff
@@ -56,6 +56,27 @@ static inline void eui48_to_eui64(eui64_t *eui64, const eui48_t *addr)
...
+{
+    /* Preserve vendor id */
+    eui48->uint8[0] = addr->uint8[0];
+    eui48->uint8[1] = addr->uint8[1];
+    eui48->uint8[2] = addr->uint8[2];
+
+    /* Use most volatile bits */
+    eui48->uint8[3] = addr->uint8[5];
+    eui48->uint8[4] = addr->uint8[6];
+    eui48->uint8[5] = addr->uint8[7];
```
Comment (maribu) [inline sys/include/net/eui48.h:77]:
```suggestion
    eui48->uint8[5] = addr->uint8[7];
    eui48_set_local(eui48);
```

Comment (maribu) [reply sys/include/net/eui48.h:77]: And I think `eui48_clear_group()` should also be done.

Comment (maribu) [reply sys/include/net/eui48.h:77]: And maybe `@post` Doxygen command to document that the generated EUI48 is marked as only locally unique and unicast?

File: sys/include/net/eui48.h [hunk 2]
```diff
@@ -56,6 +56,27 @@ static inline void eui48_to_eui64(eui64_t *eui64, const eui48_t *addr)
...
+ * @brief   Generates an EUI-48 from a 64-bit device address
+ *
+ * @warning The resulting EUI-48 is not guaranteed to be unique.
+ *
+ * @param[out] eui48    the resulting EUI-48.
+ * @param[in]  addr     a 64-bit device address
+ */
+static inline void eui64_to_eui48(eui48_t *eui48, const eui64_t *addr)
+{
+    /* Preserve vendor id */
```
Comment (miri64) [inline sys/include/net/eui48.h:69]: Sorry for stepping in late, but shouldn't it be tested somewhere that the middle bytes of the EUI-64 are `ff:fe`? Otherwise, the EUI-64 isn't convertible to an EUI-48 AFAIK.

File: sys/include/net/eui48.h (hunk 1 above)
Comment (benpicco) [reply sys/include/net/eui48.h:77]: Hm, `eui48_to_eui64()` doesn't modify those bits either.
In the end it would be used as a normal 'raw' EUI-48, there the users already take care of setting / clearing those bits if needed.

File: sys/include/net/eui48.h (hunk 2 above)
Comment (miri64) [reply sys/include/net/eui48.h:69]: See https://www.avrfreaks.net/comment/2614676#comment-2614676

Comment (maribu) [reply sys/include/net/eui48.h:69]: Isn't that conversion deprecated anyway? IMO, we should just mark the generated EUI48 as only locally unique. That would be no harm even if the middle bytes aren't `ff:fe`, or am I missing something?

Comment (benpicco) [reply sys/include/net/eui48.h:69]: Well it looks like people will do this regardless, so we might as well give them a convenience function to do it in the most effective way :wink: 

https://github.com/RIOT-OS/RIOT/pull/15083#issuecomment-698579403

There is a warning that the resulting EUI is not globally unique, still better than relying on luid.

Comment (miri64) [reply sys/include/net/eui48.h:69]: > Isn't that conversion deprecated anyway?

Huh the up-conversion? Do you have a source for that?

Comment (miri64) [reply sys/include/net/eui48.h:69]: > There is a warning that the resulting EUI is not globally unique, still better than relying on luid.

Mh... I'm not convinced. There is still a chance that this way, the generated EUI-64 for the IPv6 address from the EUI-48 is the same as the original EUI-64, resulting in two identical addresses for the device. Rather using a completely different address seems more reliable (maybe using the original EUI-64 as additional "entropy")

Comment (maribu) [reply sys/include/net/eui48.h:69]: Wouldn't just marking the EUI-48 as only locally unique already result in different IPv6 addresses, assuming the EUI-64 was globally unique?

Comment (maribu) [reply sys/include/net/eui48.h:69]: > > Isn't that conversion deprecated anyway?
> 
> Huh the up-conversion? Do you have a source for that?

See https://standards.ieee.org/content/dam/ieee-standards/standards/web/documents/tutorials/eui.pdf:

> #### Mapping an EUI-48 to an EUI-64
> Mapping an EUI-48 to an EUI-64 is deprecated. The mapping is described here for historical reasons.

To my understanding, mapping is bidirectional - so the deprecation is also for both directions. So the conversion is to my understanding on the "dark side of the force" regardless of the `ff:fe` being present.

Comment (miri64) [reply sys/include/net/eui48.h:69]: > Wouldn't just marking the EUI-48 as only locally unique already result in different IPv6 addresses, assuming the EUI-64 was globally unique?

That's true... Mh...

Comment (miri64) [reply sys/include/net/eui48.h:69]: Ok, then NVM (still want a source for the deprecation of the up-conversion though :P)

File: sys/include/net/eui48.h (hunk 1 above)
Comment (maribu) [reply sys/include/net/eui48.h:77]: Indeed, the group bit should already be cleared. But IMO a globally unique EUI-64 should still be turned into a locally unique EUI-48 for the reasons mentioned in the discussion below.

File: sys/include/net/eui48.h (hunk 2 above)
Comment (maribu) [reply sys/include/net/eui48.h:69]: > Ok, then NVM (still want a source for the deprecation of the up-conversion though :P)

See https://github.com/RIOT-OS/RIOT/pull/15096#discussion_r496584515 ;-)

Comment (miri64) [reply sys/include/net/eui48.h:69]: Oops, somehow missed that. Sorry.

Comment (maribu) [reply sys/include/net/eui48.h:69]: > Oops, somehow missed that. Sorry.

Happens to me all the time. I also often review out-of date code in github until I notice a tiny red warning at the top indicating the code displayed is out of date. I think we should blame GitHub for not refreshing properly :-D

File: sys/include/net/eui48.h [hunk 3]
```diff
@@ -56,6 +56,30 @@ static inline void eui48_to_eui64(eui64_t *eui64, const eui48_t *addr)
     eui64->uint8[7] = addr->uint8[5];
 }
 
+/**
+ * @brief   Generates an EUI-48 from a 64-bit device address
+ *
+ * @warning The resulting EUI-48 is not guaranteed to be unique.
```
Comment (maribu) [inline sys/include/net/eui48.h:62]:
```suggestion
 * @warning The resulting EUI-48 is not guaranteed to be unique
 *          and, hence, marked as only locally unique.
```

**Review by maribu — APPROVED** [review]
ACK. Please also squash the suggested documentation clarification (feel free to rephrase) right in.

File: sys/include/net/eui48.h [hunk 4]
```diff
@@ -36,6 +36,49 @@ typedef struct {
...
+ *
+ * @see IEEE 802-2001 section 9.2
+ *
+ * @param   addr    ethernet address
+ */
+static inline void eui48_clear_group(eui48_t *addr)
+{
+    addr->uint8[0] &= ~EUI48_GROUP_FLAG;
+}
+
```
Comment (maribu) [inline sys/include/net/eui48.h:81]: Could you split out the code reordering into a separate commit?

Comment (benpicco) [reply sys/include/net/eui48.h:81]: Ok

**Review by maribu — APPROVED** [review]
ACK
